[CSL-1533] Use BlkSemaphore as the only lock

gromakovsky opened the pull request:
1. `BlkSemaphore` was moved into `infra` package, because it's needed for e. g. `txp`. It required some refactoring (mostly imports fixes).
2. Txp now uses `BlkSemaphore` to ensure correctness of concurrent modifications. As a consequence, it no longer needs to compare tips.
3. Update system now uses `BlkSemaphore` for synchronization as well, there is no other lock anymore.

I would like to also do more cleaning: not store tips in mempools, remove lock from delegation, maybe something else.

kantp commented on txp/Pos/Txp/MemState/Class.hs:
```diff
@@ -100,7 +93,7 @@ modifyTxpLocalData reason f =
+        (res, logMetricsRelease) <- do
             timeEndWait <- currentTime
             liftIO . usingLoggerName lname $
                 txpMetricsAcquire (timeEndWait - timeBeginWait)
```
We should move the code that checks how long we wait for the lock to the place where the lock is now taken -- otherwise, we'll always falsely report that we don't have to wait for the lock at all.
I'd rather not lose that information, in particular since we will increase usage of the `BlkSemaphore` with this PR.

int-index replied:
Hm, would it make sense to add move these metrics into the definition of `withBlkSemaphore`?

dcoutts replied:
Yes, so we have agreed that we will extend `withBlkSemaphore` with all the nice logging and metrics that we currently have here for the tx lock processing. It's ok to have the intermediate state where this PR is merged to master, so long as we do follow-up with the additions to `withBlkSemaphore`.

kantp replied:
I agree.

dcoutts commented on infra/Pos/Infra/Semaphore.hs:
```diff
@@ -0,0 +1,80 @@
+    { unBlkSemaphore :: MVar HeaderHash
+    }
+
+-- TODO: use 'withMVar' instead ('MonadThrow'-based).
```
I'm not sure I understand this TODO, can you explain?

`withMVar` does essentially the same using a `bracketOnError` pattern.

gromakovsky replied:
Ideally I would like to use a library function instead of defining it manually. There is `withMVar` in `base` which works in `IO` only, so it's not suitable. There is `withMVar` in `lifted-base`, but it's based on `MonadBaseControl IO`, while majority of our code uses `MonadMask` for these purposes. So I would like to define `withMVar` with `MonadMask` constraint somewhere in `Pos.Util` and use it here. I also need this version of `withMVar` for another issue I am working on.
I can also use `MonadBaseControl IO` version, but anyway, I want this logic with bracket to be defined in a more common place.
Also `withMVar` is defined slightly differently in `base`:
```
withMVar :: MVar a -> (a -> IO b) -> IO b
withMVar m io =
  mask $ \restore -> do
    a <- takeMVar m
    b <- restore (io a) `onException` putMVar m a
    putMVar m a
    return b
```

I believe both versions are correct, but there are many corner cases and I would rather copy that version and trust it instead of using slightly different manually defined one.

gromakovsky replied:
Oh, btw, I confused `withMVar` and `modifyMVar`, but it doesn't really matter.

dcoutts commented on node/src/Pos/Explorer/Txp/Local.hs:
```diff
@@ -91,9 +93,14 @@ instance MonadTxExtraRead EProcessTxMode where
+    => (TxId, TxAux) -> m (Either ToilVerFailure ())
+eTxProcessTransaction itw = withBlkSemaphoreIgnoreTip $ eTxProcessTransactionNoLock itw
+
+eTxProcessTransactionNoLock
```
For cases like this, where we have bits of code that assume that the lock is held, we could use a typing trick to make sure that these functions are only called when the lock is indeed held. In this specific case it is obvious, but there are now lots of other places in the code elsewhere that are only safe to call when the lock is held.

The `withBlkSemaphore` could provide the proxy to the inner action, and these actions that need to be called with the block held could require it in their type. It'd be a good sanity check and would enforce the pattern we have in various places of `blah` and `blahNoLock`.

gromakovsky replied:
Good idea, I like it, but will leave it for the next PR :)

dcoutts commented on update/Pos/Update/Logic/Local.hs:
```diff
@@ -251,11 +256,10 @@ processVote vote = processSkeleton $ UpdatePayload Nothing [vote]
-        tip <- DB.getTip
-        stateVar <- mvState <$> views (lensOf @UpdateContext) ucMemState
-        atomically . writeTVar stateVar =<< usNormalizeDo (Just tip) Nothing
+usNormalize = do
```
Here's another case where the typing trick to enforce that we hold the lock would be useful, rather than just having it as a comment. It's easy to make mistakes with this or in later refactoring.

dcoutts commented on update/Pos/Update/Logic/Local.hs:
```diff
@@ -285,16 +289,19 @@ usNormalizeDo tip slot = do
+processNewSlot :: (USLocalLogicModeWithLock ctx m) => SlotId -> m ()
+processNewSlot slotId = withUSLock $ processNewSlotNoLock slotId
+
+processNewSlotNoLock :: (USLocalLogicMode ctx m) => SlotId -> m ()
```
Here's another instance of the same pattern.

dcoutts commented on node/src/Pos/Explorer/Txp/Local.hs:
```diff
@@ -95,17 +94,15 @@ eTxProcessTransaction itw = withBlkSemaphoreIgnoreTip $ eTxProcessTransactionNoL
+    let UnsafeTx {..} = taTx txAux
+    -- Note: this function is assumed to be run while we own
+    -- 'BlkSemaphore', so we can be sure GState won't change while we
+    -- are doing these operations.
```
Another good example where a type would help. Types win over comments. :-)

(But this comment is useful too since it explains why it is correct)

gromakovsky commented on infra/Pos/Infra/Semaphore.hs:
```diff
@@ -0,0 +1,80 @@
+    { unBlkSemaphore :: MVar HeaderHash
+    }
+
+-- TODO: use 'withMVar' instead ('MonadThrow'-based).
```
I think I will do it in this PR.
It's not used, because it's not defined 😉 
As I wrote below, there is `IO` version in `base` and `MonadBaseControl IO` version in `lifted-base`, but no `MonadMask` version.

gromakovsky commented:
I have some updates.
1. I got rid of TODO in `Semaphore.hs`, now there are utility functions in `Pos.Util.Concurrent` and functions in `Semaphore.hs` are just wrappers for them. Also I slightly refactored and renamed these wrappers to be more consistent with the naming used in `Control.Concurrent.MVar`.
2. I considered a remark on exception safety from the write-up by @int-index (last paragraph [here](https://github.com/input-output-hk/cardano-sl/blob/int-index/csl-1533/csl-1533-overview.md#semi-formal-proof-of-correctness)). That's indeed the problem.
It would be good to assume that with new locking machinery our mempool always corresponds to the same tip (and the same GState) as the one stored in our database (and one can think that's true, because whenever we update GState, we also normalize mempool), but unfortunately we can't rely on that, because after we update GState there can be an exception or something else can go wrong and normalization can fail. In this case we'll have a mempool which doesn't correspond to our DB.
In order to resolve this problem I reverted the change which gets rid of tip check in tx processing. We still need this check. We can guarantee that mempool corresponds to the tip from `TVar`, but can't guarantee it corresponds to the tip from DB.
Note, however, that it's different from our previous problem. Previously this check could fail while we were processing blocks and it was happening quite often. Now it can happen only if an unexpected exception occurs which probably indicates that something is bad and rejecting transactions is not our main problem. So I think it's fine and we don't need to improve anything here.
3. Related to (2): there are few more places where we store data from the DB in memory and rely on their correspondence, but don't consider exception safety. I will create another issue about it.
4. It's also important to check in-memory tip before creating a new block. It must be the same as the tip in DB. I added this check there as well, if tip differs, we use empty payload.
5. I also did similar changes in update system.
6. I decided not to touch delegation for now, because initially I didn't plan to spend much time on this issue, but have already spent quite a lot. I will create a separate issue for it. I don't want to delay merge of this PR too much.

gromakovsky commented:
A new issue about further efforts: https://issues.serokell.io/issue/CSL-1558
I tentatively assigned it to @kantp, maybe someone else will participate (there are several things to do).
A new issue about locking in delegation: https://issues.serokell.io/issue/CSL-1559
I tentatively assigned it to @volhovM.
A new issue about synchronization of in-memory data and data from the DB: https://issues.serokell.io/issue/CSL-1560
